Add document about local ssh (#3800)
gitpod-io/website: gitpod/docs/references/ides-and-editors/vscode-desktop-local-ssh.md

---
section: ides-and-editors
title: VS Code Desktop
---

<script context="module">
  export const prerender = true;
  import Keybind from "$lib/components/keybind.svelte";
</script>

# VS Code Desktop - Local SSH

We now offer a new option: `Local SSH` to access your workspaces with VS Code Desktop. This method leverages the stored Gitpod session in VS Code Desktop for authentication, providing an additional layer of security.

## Modify SSH Config File

Please be aware that to use this method, we will need to add some content to your ssh config file.

> > **🚨 Security Note:** To use `Local SSH`, **it requires to add some content to your ssh config file**. Please remember that if you ever have any security concerns, you could use other methods or choose other editors.

Reviewer: We want to get rid of other methods? Maybe don't mention it and see who come at us?

> ---

Reviewer 2: **Suggestion:** Can this page be integrated with the existing page, I'd also suggest we use the text "connecting to VS Code" rather than mentioning "Local SSH" as the name. It would be quite hard as a user to know what "Local SSH" is and/or whether the user needs to know about it or not. We have a section already for connecting, could we integrate the documentation there, rather than a separate page? 🙏 

<img width="293" alt="image" src="https://github.com/gitpod-io/website/assets/5528307/fa3d2d4d-d460-46d3-a028-40d3ad9bd629">

section: ides-and-editors
title: VS Code Desktop - Local SSH
---

<script context="module">
  export const prerender = true;
  import Keybind from "$lib/components/keybind.svelte";
</script>

# VS Code Desktop - Local SSH

> We now offer a new option: `Local SSH` to access your workspaces with VS Code Desktop. This method leverages the stored Gitpod session in VS Code Desktop for authentication, providing an additional layer of security.

Reviewer 3: suggestion: Let's make this sounds less like a changelog entry, because it will not be **new** forever 😄 

```suggestion
`Local SSH` allows users to access their workspaces from VS Code Desktop. This method utilizes the stored Gitpod session in VS Code Desktop as a form of authentication, thereby providing an additional layer of security.
```

## Modify SSH Config File

Please be aware that to use this method, modifications to your SSH config file are necessary.

> <!-- > **🚨 Security Note:** To use `Local SSH`, **it requires to add some content to your ssh config file**. Please remember that if you ever have any security concerns, you could use other methods or choose other editors. -->

Reviewer 3: Like below, I think this should be uncommented
```suggestion
> **🚨 Security Note:** To use `Local SSH`, Gitpod has to add configuration to your SSH config file. Please remember that if you have security concerns about the approach, you can use other methods or choose other editors.
```

### Why We Need to Modify Your SSH Configurations?

> The Local SSH Proxy requires changes to your SSH config file. This is because there is no official way provided by VS Code Desktop to hook the SSH authentication. Therefore, we need to adjust the SSH configuration to allow the our Gitpod Extension to use the stored Gitpod session to get the SSH password and do authentication automatically for you.

Reviewer 3: ```suggestion
The Local SSH Proxy requires changes to your SSH config file. This is because there is no official way provided by VS Code Desktop to hook into the SSH authentication. Therefore, we need to adjust the SSH configuration to allow the Gitpod Extension to use the stored Gitpod session for retrieving the SSH password and doing authentication automatically for you.
```

### What Changes Are Made?

> Here are the steps we take to modify your SSH configurations:

Reviewer 3: nitpick:

```suggestion
The following steps illustrate how we modify your SSH configurations:
```

**macOS / Linux:**

1. Creating a new file `~/.ssh/code_gitpod.d/config` and inserting our settings.

> 2. Updating your ssh config file `~/.ssh/config` to import the file created in step one. This will include grouping `Include "code_gitpod.d/config"` statement under the header `## START GITPOD INTEGRATION.`

Reviewer 3: ```suggestion
2. Updating your SSH config file `~/.ssh/config` to import the file created in step one. This will include grouping `Include "code_gitpod.d/config"` statement under the header `## START GITPOD INTEGRATION.`
```

```shell
$ cat ~/.ssh/config
## START GITPOD INTEGRATION
## This section is managed by Gitpod. Any manual changes will be lost.
Include "code_gitpod.d/config"
## END GITPOD INTEGRATION
```

```shell
$ cat ~/.ssh/code_gitpod.d/config
### This file is managed by Gitpod. Any manual changes will be lost.

Host *.vss.gitpod.io
  ...
```

![Warning on dashboard](/images/editors/local-ssh-warning-on-dashboard-dark-theme.png)

> ![Warning on dashboard](/images/editors/local-ssh-warning-on-dashboard-light-theme.png)

Reviewer 2: There are some typos in the screenshot: 

* "Common on" -> "Comment on"

Is there a related PR that I can review? I assume these are drafts? 🙏

_Warning on dashboard_

## Local SSH

### What is Local SSH?

Local SSH Proxy is a method used to connect your Gitpod workspace with VS Code Desktop. It's designed to increase convenience and maintain a secure connection between your local environment and the Gitpod workspace. Local SSH acts like a proxy that uses the stored Gitpod session in VS Code Desktop to establish an SSH tunnel to your Gitpod workspace. This eliminates the need for manually uploading SSH keys or copying the SSH password each time a connection or reconnection with SSH Gateway, making the process of connecting to your Gitpod workspace smoother.

Essentially, the Local SSH Proxy enhances the experience of using Gitpod with VS Code Desktop, making it a smoother, more integrated process. It's all about improving your development workflow, reducing the friction and allowing you to focus on your code.

![What is Local SSH](/images/editors/what-is-local-ssh-dark-theme.png)

> ![What is Local SSH](/images/editors/what-is-local-ssh-light-theme.png)

Reviewer 2: I really like that we have a diagram for this 😍 — I think something that's not super clear on the diagram is the reasoning for the Local SSH configuration, which is really so that Remote SSH can read the SSH configuration and proxy to the tunnel.

Author: diagram updated
